accepted/2021/User-Experience/Apple-Silicon-UX.md

```diff
@@ -0,0 +1,254 @@
+
+For instance:
+
+`sudo mv $(dirname $(which dotnet)) $(dirname $(which dotnet))-x64`
+
+Which version to be used must be controlled by the path environment variable.
+
+#### Design Constraints
+
+- This mixed forced architecture problem is transient in nature.
+
+- We want to avoid making major architectural changes to support this temporary design problem.
+
+- We want to keep the process parent-child relationships as consistent as possible.
+
+- We want minimal changes to the earlier releases.
+
+#### Design Experiment
+
+https://github.com/dotnet/runtime/pull/48541 represented an early design experiment to allow side-by-side macOS-x64 .NET 5 with macOS-arm64 .NET 6 to coexist.
+
+The experiment:
+
+- Assumed .NET runtime versions less than 6.0 will always require `x86_64` architecture. All other runtime versions will require `arm64` architecture on Apple Silicon.
+
+- Made the .NET 6 `dotnet` & `libhostfxr.dylib` (`hostfxr`) universal binaries. This means both native and emulated processes can perform runtime resolution to identify the correct runtime to use.
+
+- The `hostfxr` was modified to enforce the architecture version  policy above. It will return an architecture mismatch error when the runtime (by version policy above) require a different architecture than the currently running process.
+
+- `dotnet` was modified to enable switching the architecture of the current process when it saw an architecture mismatch error returned by `hostfxr`. The current process will switch architecture as needed and run with the same arguments as the new process. The switching will use `posix_spawn` with changes to `posix_spawnattr_t` to set `POSIX_SPAWN_SETEXEC` and preferred architecture via `posix_spawnattr_setbinpref_np` (see Apple Silicon man page) and setting `sysctl` `kern.curproc_arch_affinity` property appropriately.
+
+Overall the experiment demonstrated a highly workable user experience.
+
+- The experiment represents a viable design alternative. It did not support side-by-side install of multiple architectures of the same runtime version.
+
+- The implementation was incomplete. It did not automate the build process or deal with installer issues
+
+### Dotnet Publish Experience Inconsistent
+
+#### Signing
+
+- dotnet publish for Xamarin workloads handles signing automatically.
+- dotnet publish for CoreClr console apps do not handle signing automatically. The developer would need to explicitly sign after publishing for Apple Silicon.  The signing command is relatively simple `codesign -s <signature> --entitlements <app-entitlements-path> <app-path>`.
+- There is no message to the customer that the published app needs to be signed
+
+#### Dotnet Publish as Universal Binaries
+
+- This scenario is also supported for Xamarin iOS workloads. This is a slightly different scenario as Xamarin iOS apps are native binaries (not JIT binaries)
+- CoreCLR runtime apps have no mechanism to publish universal binaries.
```

Comment:
Is that the correct message if we considering coreclr JIT to be backend for Xamarin.Mac ?

Reply:
Right, we would need to make this work as part of adding support for coreclr w/ Xamarin.Mac.